src/cmd/vendor.rs

```diff
@@ -0,0 +1,645 @@
+                        &rt,
+                        git,
+                        vendor_package,
+                        patch_link,
+                        dep_path.clone(),
+                        matches,
+                    )
+                })
+            }
+
+            Some(("patch", matches)) => {
+                // Apply patches
+                let mut num_patches = 0;
+                patch_links.clone().into_iter().try_for_each(|patch_link| {
+                    apply_patches(&rt, git, vendor_package.name.clone(), patch_link)
+                        .map(|num| num_patches += num)
+                })?;
+
+                // Commit applied patches to clean working tree
+                if num_patches > 0 {
+                    rt.block_on(git.add_all())?;
+                    rt.block_on(git.commit(Some("pre-patch")))?;
+                }
+
+                // Generate patch
+                patch_links.clone().into_iter().try_for_each( |patch_link| {
+                    match patch_link.patch_dir.clone() {
+                        Some(patch_dir) => {
+                            if matches.is_present("plain") {
+                                let get_diff = diff(&rt,
+                                                    git,
+                                                    vendor_package,
+                                                    patch_link,
+                                                    dep_path.clone())
+                                            .map_err(|cause| Error::chain("Failed to get diff.", cause))?;
+                                gen_plain_patch(get_diff, patch_dir, false)
+                            } else {
+                                gen_format_patch(&rt, sess, git, patch_link, vendor_package.target_dir.clone())
+                            }
+                        },
+                        None => {
+                            warnln!("No patch directory specified for package {}, mapping {} => {}. Skipping patch generation.", vendor_package.name.clone(), patch_link.from_prefix.to_str().unwrap(), patch_link.to_prefix.to_str().unwrap());
+                            Ok(())
+                        },
+                    }
+                })
+            }
+            _ => Ok(()),
+        }?;
+    }
+
+    Ok(())
+}
+
+/// initialize the external dependency
+pub fn init(
+    rt: &Runtime,
+    git: Git,
+    vendor_package: &config::VendorPackage,
+    patch_link: PatchLink,
+    dep_path: impl AsRef<Path>,
+    matches: &ArgMatches,
+) -> Result<()> {
+    // import necessary files from upstream, apply patches
+    let dep_path = dep_path.as_ref();
+
+    // Make sure the target directory actually exists
+    let link_to = patch_link
+        .to_prefix
+        .clone()
+        .prefix_paths(&vendor_package.target_dir);
+    let link_from = patch_link.from_prefix.clone().prefix_paths(dep_path);
+    std::fs::create_dir_all(&link_to.parent().unwrap())?;
```

Comment:
Again, we should print some nicer error message if directory creation fails (e.g. due to issues with write permissions).